File: Cmdlets/GetCurrencyInfoCommand.vb
Imports System.Collections.Generic
Imports System.Management.Automation

<Cmdlet("Get", "CurrencyInfo")>
Public Class GetCurrencyInfoCommand
	Inherits Cmdlet

	<Parameter>
	Public Property CurrencyRegions() As SwitchParameter

	<Parameter>
	Public Property ISOCurrencySymbol() As String

	Protected Overrides Sub BeginProcessing()

		Dim ISOCodeSpecified As Boolean = Not String.IsNullOrEmpty(Me.ISOCurrencySymbol)

		If Me.CurrencyRegions.IsPresent Then

			If ISOCodeSpecified Then

				Dim RegionalCurrencies = Me.GetSpecificCurrencyRegions(Me.ISOCurrencySymbol)

				Me.WriteObject(RegionalCurrencies, True)
			Else

				Dim RegionalCurrencies = Me.GetAllCurrencyRegions()

				Me.WriteObject(RegionalCurrencies, True)
			End If
		Else

			If ISOCodeSpecified Then

				Dim Currencies = Me.GetSpecificCurrency(Me.ISOCurrencySymbol)

				Me.WriteObject(Currencies, True)
			Else

				Dim Currencies = Me.GetAllCurrencies()

				Me.WriteObject(Currencies, True)
			End If
		End If

	End Sub

	Private Function GetAllCurrencyRegions() As List(Of CurrencyRegions)
		Dim Regions = GetAllRegions()

		Dim RegionalCurrencies = New List(Of CurrencyRegions)

		For Each Region In Regions

			Dim Currency = New TIKSN.Finance.CurrencyInfo(Region)

			If RegionalCurrencies.Any(Function(RC) RC.Currency = Currency) Then

				Dim RegionalCurrency = RegionalCurrencies.Single(Function(RC) RC.Currency = Currency)

				RegionalCurrency.Regions.Add(Region)
			Else
				Dim RegionalCurrency = New CurrencyRegions

				RegionalCurrency.Currency = Currency
				RegionalCurrency.Regions.Add(Region)

				RegionalCurrencies.Add(RegionalCurrency)
			End If
		Next

		Return RegionalCurrencies
	End Function

	Private Function GetSpecificCurrencyRegions(ISOCode As String) As System.Collections.Generic.List(Of CurrencyRegion)

		Dim Regions = GetAllRegions()

		Dim Result = New System.Collections.Generic.List(Of CurrencyRegion)

		For Each Region In Regions

			If String.Compare(Region.ISOCurrencySymbol, ISOCode, System.StringComparison.InvariantCultureIgnoreCase) = 0 Then

				Result.Add(New CurrencyRegion(Region))
			End If
		Next

		Return Result
	End Function

	Private Function GetAllCurrencies() As HashSet(Of Finance.CurrencyInfo)
		Dim Regions = GetAllRegions()

		Dim Currencies As New HashSet(Of Finance.CurrencyInfo)

		For Each Region In Regions
			Currencies.Add(New TIKSN.Finance.CurrencyInfo(Region))
		Next

		Return Currencies
	End Function

	Private Function GetSpecificCurrency(ISOCode As String) As System.Collections.Generic.IEnumerable(Of TIKSN.Finance.CurrencyInfo)
		Dim Regions = GetAllRegions()

		Dim Currencies As New List(Of Finance.CurrencyInfo)

		For Each Region In Regions

			If String.Compare(Region.ISOCurrencySymbol, ISOCode, System.StringComparison.InvariantCultureIgnoreCase) = 0 Then

				Currencies.Add(New TIKSN.Finance.CurrencyInfo(Region))
			End If
		Next

		Return Currencies
	End Function
End Class

File: C#/Cmdlets/Cmdlets/GetCurrencyInfoCommand.cs
using System.Collections.Generic;
using System.Linq;
using System.Management.Automation;
using TIKSN.Cmdlets.Helpers;

namespace TIKSN.Cmdlets
{
    [Cmdlet("Get", "CurrencyInfo")]
    public class GetCurrencyInfoCommand : Cmdlet
    {
        [Parameter()]
        public SwitchParameter CurrencyRegions { get; set; }

        [Parameter()]
        public string ISOCurrencySymbol { get; set; }

        protected override void BeginProcessing()
        {
            bool ISOCodeSpecified = !string.IsNullOrEmpty(this.ISOCurrencySymbol);

            if (this.CurrencyRegions.IsPresent)
            {
                if (ISOCodeSpecified)
                {
                    var RegionalCurrencies = this.GetSpecificCurrencyRegions(this.ISOCurrencySymbol);

                    this.WriteObject(RegionalCurrencies, true);
                }
                else
                {
                    var RegionalCurrencies = this.GetAllCurrencyRegions();

                    this.WriteObject(RegionalCurrencies, true);
                }
            }
            else
            {
                if (ISOCodeSpecified)
                {
                    var Currencies = this.GetSpecificCurrency(this.ISOCurrencySymbol);

                    this.WriteObject(Currencies, true);
                }
                else
                {
                    var Currencies = this.GetAllCurrencies();

                    this.WriteObject(Currencies, true);
                }
            }
        }

        private HashSet<Finance.CurrencyInfo> GetAllCurrencies()
        {
            var Regions = RegionHelper.GetAllRegions();

            HashSet<Finance.CurrencyInfo> Currencies = new HashSet<Finance.CurrencyInfo>();

            foreach (var Region in Regions)
            {
                Currencies.Add(new TIKSN.Finance.CurrencyInfo(Region));
            }

            return Currencies;
        }

        private List<CurrencyRegions> GetAllCurrencyRegions()
        {
            var Regions = RegionHelper.GetAllRegions();

            var RegionalCurrencies = new List<CurrencyRegions>();

            foreach (var Region in Regions)
            {
                var Currency = new TIKSN.Finance.CurrencyInfo(Region);

                if (RegionalCurrencies.Any(RC => RC.Currency == Currency))
                {
                    var RegionalCurrency = RegionalCurrencies.Single(RC => RC.Currency == Currency);

                    RegionalCurrency.Regions.Add(Region);
                }
                else
                {
                    var RegionalCurrency = new CurrencyRegions();

                    RegionalCurrency.Currency = Currency;
                    RegionalCurrency.Regions.Add(Region);

                    RegionalCurrencies.Add(RegionalCurrency);
                }
            }

            return RegionalCurrencies;
        }

        private System.Collections.Generic.IEnumerable<TIKSN.Finance.CurrencyInfo> GetSpecificCurrency(string ISOCode)
        {
            var Regions = RegionHelper.GetAllRegions();

            List<Finance.CurrencyInfo> Currencies = new List<Finance.CurrencyInfo>();

            foreach (var Region in Regions)
            {
                if (string.Compare(Region.ISOCurrencySymbol, ISOCode, System.StringComparison.InvariantCultureIgnoreCase) == 0)
                {
                    Currencies.Add(new TIKSN.Finance.CurrencyInfo(Region));
                }
            }

            return Currencies;
        }

        private System.Collections.Generic.List<CurrencyRegion> GetSpecificCurrencyRegions(string ISOCode)
        {
            var Regions = RegionHelper.GetAllRegions();

            var Result = new System.Collections.Generic.List<CurrencyRegion>();

            foreach (var Region in Regions)
            {
                if (string.Compare(Region.ISOCurrencySymbol, ISOCode, System.StringComparison.InvariantCultureIgnoreCase) == 0)
                {
                    Result.Add(new CurrencyRegion(Region));
                }
            }

            return Result;
        }
    }
}